### PR TITLE
[wip] interactive: App - added ability to apply passes 

### PR DESCRIPTION
**Pull request by dshaaban01**
Using Textual Python, I have created a GUI/app.This app allows you to write/paste xDSL IR into the "Input" TextArea, select a pass (or multiple passes), which are applied onto the input IR and are displayed in the "Output" TextArea. The selected passes are made visible to the user in the "Selected passes/query" Box. Buttons have also been added to make the GUI more usable. "Clear Input" Button, "Copy Output" Button, "Copy Query" Button (copies the query you have built via selecting passes) and "Clear Passes" Button (removes all the selected passes from the IR). Tests have also been written.

This app is still under construction, this is the second iteration towards building the full tool.

### CODE REVIEW

File: tests/interactive/test_app.py [hunk 1]
```diff
@@ -1,22 +1,69 @@
...
+        pipeline.apply(ctx, module)
+
+        res = module
+    except Exception as e:
+        res = e
+
+    return res
+
+
+def transform_input_return_str(res: ModuleOp | Exception) -> str:
```
Comment (superlopuh) [inline tests/interactive/test_app.py:43]: why move this function here?

File: tests/interactive/test_app.py [hunk 2]
```diff
@@ -69,11 +116,170 @@ async def test_input():
...
+                test_module, ModuleOp
+            ):
+                assert app.current_module.is_structurally_equivalent(test_module)
+
+                # assert that the input and output text area's are equal
+            await pilot.pause()
+            assert app.output_text_area.text == test_output_area_str
+
+
+# @pytest.mark.asyncio()
```
Comment (superlopuh) [inline tests/interactive/test_app.py:221]: can you please remove the commented code? Or uncomment it?

File: xdsl/interactive/app.py [hunk 3]
```diff
@@ -52,41 +58,86 @@ class InputApp(App[None]):
...
+    query_label = Label("", id="selected_passes_label")
+
     def compose(self) -> ComposeResult:
         """
         Creates the required widgets, events, etc.
         Get the list of xDSL passes, add them to an array in "Selection" format (so it can be added to a Selection List)
         and sort the list in alphabetical order.
         """
 
+        # construct the seleciton list containing all the passes and the label displaying the selected passes
```
Comment (webmiche) [inline xdsl/interactive/app.py:85]:
```suggestion
        # construct the selection list containing all the passes and the label displaying the selected passes
```

File: xdsl/interactive/app.py [hunk 4]
```diff
@@ -121,6 +173,17 @@ def on_mount(self) -> None:
 
         self.query_one("#input_container").border_title = "Input xDSL IR"
         self.query_one("#output_container").border_title = "Output xDSL IR"
+        self.query_one(
+            "#passes_selection_list"
+        ).border_title = "Choose a pass or multiple passes to be applied."
+        self.query_one("#selected_passes").border_title = "Selected passes/query"
+
+        # aids in the construction of the seleciton list containing all the passes
```
Comment (webmiche) [inline xdsl/interactive/app.py:181]:
```suggestion
        # aids in the construction of the selection list containing all the passes
```

**Review by webmiche — COMMENTED** [review]
I think this PR does a bit too much. There are four buttons added, and all the logic for selecting and executing passes. I think, all of these could be individual PRs probably.

**Pull request closed without merging**